src/distributed_copy.c

```diff
@@ -0,0 +1,857 @@
+ * We can not use connection cache from connection.c used by GteConnection because
+ * we need to establish multiple connections with each nodes: one connection per shard
+ */
+static HTAB *
+CreateShardToConnectionHash()
+{
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(ShardId);
+	info.entrysize = sizeof(ShardConnections);
+	info.hash = shard_id_hash_fn;
+
+	return hash_create("shardToConn", MAX_SHARDS, &info, HASH_ELEM | HASH_FUNCTION);
+}
+
+/* 
+ * Reconstruct text of COPY statement based on CopyStmt for the particular shard
+ */
+static char const *
+ConstructCopyStatement(CopyStmt *copyStatement, ShardId shardId)
+{
+	StringInfo buf = makeStringInfo();
+	ListCell *cell = NULL;
+	char sep = '\0';
+	char const *qualifiedName = quote_qualified_identifier(
+		copyStatement->relation->schemaname,
+		copyStatement->relation->relname);
+	appendStringInfo(buf, "COPY %s_%ld ", qualifiedName, (long) shardId);
+	if (copyStatement->attlist)
+	{
+		sep = '(';
+		foreach(cell, copyStatement->attlist)
+		{
+			appendStringInfoChar(buf, sep);
+			appendStringInfoString(buf, strVal(lfirst(cell)));
+			sep = ',';
+		}
+		appendStringInfoChar(buf, ')');
+	}
+	appendStringInfoString(buf, "FROM STDIN");
+
+	if (copyStatement->options)
+	{
+		appendStringInfoString(buf, " WITH ");
+		sep = '(';
+		foreach(cell, copyStatement->options)
+		{
+			DefElem *def = (DefElem *) lfirst(cell);
+			appendStringInfo(buf, "%c%s '%s'", sep, def->defname, defGetString(def));
+			sep = ',';
+		}
+		appendStringInfoChar(buf, ')');
+	}
+	return buf->data;
+}
+
+/*
+ * Send PQputCopyEnd command to the client and check result (status of 
+ * COPY command completion)
+ */
+static bool
+PgCopyEnd(PGconn *con, char const* msg)
+{
+	PGresult *result = NULL;
+	if (PQputCopyEnd(con, msg) == 1)
+	{
+		while ((result = PQgetResult(con)) != NULL)
+		{
+			if (PQresultStatus(result) != PGRES_COMMAND_OK)
+			{
+				if (msg == NULL)
+				{
+					ReportRemoteError(con, result);
+				}
+				return false;
+			}
+			PQclear(result);
+		}
+		return true;
+	}
+	else
+	{
+		ReportRemoteError(con, NULL);
+	}
+	return false;
+}
+
+/*
+ * End copy and prepare transaction.
+ * This function is applied for each shard placement unless some error happen.
+ * Status of this function is stored in ShardConnections::status field
+ */
+static ShardId
+PgCopyPrepareTransaction(List* shardConnections)
+{
+	ListCell *listCell;
+	int i = 0;
+	PgShardTransactionManager const *tmgr =
+		&PgShardTransManagerImpl[PgShardCurrTransManager];
+	foreach (listCell, shardConnections)
+	{
+		ShardConnections* shardConn = (ShardConnections*)lfirst(listCell);
+		for (i = 0; i < shardConn->replicaCount; i++) 
+		{
+			PGconn* conn = shardConn->placements[i].conn;
+			shardConn->placements[i].copied = true;
+			if (PgCopyEnd(conn, NULL) && tmgr->Prepare(conn))
+			{
+				shardConn->placements[i].prepared = true;
+			}
+			else
+			{
+				return shardConn->shardId;
```

Comment:
I think it would be nicer to just do an ereport(ERROR here and return void, effectively moving the message from the end of PgShardCopyFrom to here.